File: Data Structures & Algorithms/eating-bananas/submission-6.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        res = r

        while l <= r:
            rate = l + (r - l) // 2
            hours = 0
            for val in piles:
                hours += math.ceil(val / rate)
            
            if hours <= h:
                res = min(res, rate)
                r = rate - 1
            else:
                l = rate + 1
        return res

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-6.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0;
        int minPrice = prices[0];
        for(int price: prices) {
            minPrice = min(minPrice, price);
            res = max(res, price - minPrice);
        }
        return res;
    }
};
